### ISSUE TITLE
FastAPI always returns content, even if 204 no content status code is set.

### ISSUE THREAD

**Issue by aviramha** [tiangolo/fastapi#2832]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.
### Example

```Python
from fastapi import FastAPI
from starlette.status import HTTP_204_NO_CONTENT
app = FastAPI()


@app.get("/", status_code=HTTP_204_NO_CONTENT)
def read_root():
    print("d")
```

### Description

Send a request with any HTTP Client, observe the traffic using tcpdump/wireshark - you'll see that you get null as a response (content-length will be 4). This is extremely notable as this caused this issue https://github.com/encode/httpx/issues/1474 (The client doesn't expect any data in that case)
I would expect FastAPI to not return null when response is None.

### Environment

* OS: macOS
* FastAPI Version [e.g. 0.3.0]: 0.63.0
* Python version: 0.63.0

**Comment by Kludex** [comment]
Can you try to run with httptools instead of h11?

**Comment by aviramha** [comment]
> Can you try to run with httptools instead of h11? 

Client or server ?

**Comment by Kludex** [comment]
unicorn --http httptools

**Comment by falkben** [comment]
`httptools` will return `null` and `h11` will return empty but raises an exception if you don't explicitly return a Response but declare the status code to be `204`.

you have to explicitly return an empty Response instead of implicitly returning `None`, which is what happens when you don't return anything.

```python
return Response(status_code=HTTP_204_NO_CONTENT)
```

also, see: https://github.com/tiangolo/fastapi/issues/2253

**Comment by falkben** [comment]
You could instead set the response class for that path operator to be `Response` instead of the default `JSONResponse`.  Then it won't convert `None` into JSON.  So w/ your example above it would be:

```python
@app.get("/", status_code=HTTP_204_NO_CONTENT, response_class=Response)
def read_root():
    print("d")
```

which correctly returns `b""`.

note, if you are doing this on multiple endpoints, you can set the default response class on your router, or even your entire app. https://fastapi.tiangolo.com/advanced/custom-response/#default-response-class

**Comment by aalifadv** [comment]
@falkben the refernced PR #2833 is proposing to do what you mentioned automatically. with that PR in mind, I think it is a question of, should it be handled automatically by the framework or explicity set by the developer. Curious what others think.

**Comment by aviramha** [comment]
Thanks @falkben. 
I found this solution also but I believe it should be handled by the framework. It's fairly simple to implement hence the PR I submitted. Returning 204 with data is invalid according to the standards so FastAPI should detect such scenario and handle it accordingly IMO

**Comment by ycd** [comment]
I think this is valid for **205** (_Reset Content_) too or not?

**Comment by mikegrima** [comment]
This can also be reproduced if you raise an exception: `HTTPException(status_code=204)`

**Comment by declaresub** [comment]
I ran across this issue in the course of evaluating FastAPI for a large project.

I am undecided as to whether this is a bug or a gotcha in starlette and fastapi.  The ASGI specification states that servers must not begin to send a response until at least one 'http.response.body' event is sent, so starlette/fastapi needs to send such an event. 

Right now I think that the solution would be for starlette to add an EmptyResponse class to be used with 1xx, 204, 304, and 205 (which can have an empty content body according to RFC 7231) responses.  This makes the special cases explicit, and keeps the special-case code in one place.

I think that it is ultimately the responsibility of the server  (e.g. uvicorn) to send 1xx, 204, 304 responses without a message body.  So perhaps a uvicorn bug report needs to be written.

**Comment by Kludex** [comment]
It's a bug here.

**Comment by declaresub** [comment]
I've suggested that Starlette add an EmptyResponse class (https://github.com/encode/starlette/issues/1178) that does the right thing for status codes 1xx, 204, 205, 304.

**Comment by marceloalvescl** [comment]
For anyone who may comes here for a quick solution, here is how I solved it.
In order to make it work, I've returned Response(status_code=status.HTTP_204_NO_CONTENT)

(class starlette.responses.Response)

**Comment by wombatonfire** [comment]
While I think it's a tempting idea to have such a smart framework that handles all possible use cases, it's hardly achievable.

FastAPI does not **always** return content, as stated in the issue title, it just does it by default. The default `JSONResponse`, which is used implicitly, always has a body. This may look strange, of cause, but I think it should be seen as an implementation detail.

What's important, is that you are not forced to use the defaults. Response class can be configured globally, or individually per path operation function. A developer has full control. The [solution](https://github.com/tiangolo/fastapi/issues/2832#issuecomment-781595594) described by @falkben is the right way to go. You know that there should be no body, and thus you replace default `JSONResponse` with `Response`.

I don't think this is a bug that needs changes in the codebase, but some improvements (examples?) in the documentation, describing explicitly that default `JSONResponse` always has a body, and that it may not always be what you want (204 case), may help new users not familiar with the framework.

**Comment by TBBle** [comment]
I don't think we need to handle _all possible_ use-cases.

This bug is about the specific case where the route is defined with (default) `status_code=HTTP_204_NO_CONTENT`, because if you do not also override other defaults (i.e. the extra setting @falkben describes, or expressly returning a `Reponse` in your handler), FastAPI generates an _invalid_ HTTP responses.

Documenting

> If you set `status_code=HTTP_204_NO_CONTENT`, you must also set `response_class=Response` or some other response_class that does the right thing

is less-kind to the library user than

> If you set `status_code=HTTP_204_NO_CONTENT`, the default `response_class` will be `Response`, rather than `JSONResponse`

as in the latter case, the user specifying their status code gets valid behaviour for that status code out-of-the-box, and retains exactly the same amount of control for the user to override the defaults that exists now. Overriding that `response_class` default stops being _mandatory_, and natural things like not returning a value from a 204-typed API remain explicitly natural. And if the user decides that returning a body in their 204 response is actually desired, then that's still possible explicitly.

This also means the list of the HTTP status codes for which no response body is allowed by the standard is kept in the library code, rather than in the documentation _and_ the user's code-base.

I liked the idea of `EmptyResponse` to be the default `response_class` when a `status_code` which disallows response bodies is specified, but it seems Starlette upstream is going in a different direction there, simply explicitly excluding the `Content-Length` header in 204 (and similar) responses, which has the effect that you cannot provide a message body in a 204 response. What you get if you try (i.e the example in this issue) is a 204 response with no body, plus a bunch of garbage data on the wire for which [interpretations and handling varies](https://github.com/httpwg/http-core/issues/26#issue-231080623). I've no idea what the various Python http libraries will make of this, but I suspect it won't magically fix this problem by discarding any unintentionally-produced body data.

**Comment by wombatonfire** [comment]
> I liked the idea of `EmptyResponse` to be the default `response_class` when a `status_code` which disallows response bodies is specified, but it seems Starlette upstream is going in a different direction there, simply explicitly excluding the `Content-Length` header in 204 (and similar) responses, which has the effect that you cannot provide a message body in a 204 response.

`EmptyResponse` didn't look like a good idea to me (I've seen that issue). Why would you want to add a new response class when an already existing `Response` can perfectly be empty? When you don't pass content to it, it will return `b""`.

The attempt to fix headers in Starlette makes much more sense, since Starlette is a HTTP library, so it should be compliant with the HTTP standard.

> What you get if you try (i.e the example in this issue) is a 204 response with no body, plus a bunch of garbage data on the wire for which [interpretations and handling varies](https://github.com/httpwg/http-core/issues/26#issue-231080623).

So the scenario is that someone constructs a 204 response, asking for no content, and at the same time explicitly adding content? From a HTTP library this doesn't make much sense. QA working to intentionally break things? :)

Don't get me wrong, I do understand where you are coming from. From the downstream framework perspective, you would expect the upstream HTTP library to be standards-compatible and to not allow the invalid responses to be constructed. But the scenario you have described illustrates the **intention** to construct the invalid response. When you use 204, you know there should be no content, so you shouldn't put any content to the response.

I don't know, maybe Starlette, after fixing headers issue, could also strip the content from the response in case it was provided with the status codes that don't allow it? Just an idea, I didn't dig deeper into the problem, since I'm fine explicitly specifying correct response class in FastAPI. AFAIK, they are aware of the special cases with 204 and some other response codes, and if I understood the discussion correctly, they were going to have a look at those cases after implementing headers fixes.

**Comment by declaresub** [comment]
The problem is that a 204 response must contain neither a message body nor a Content-Length header.  A JSONResponse object does not generate such a message when the status code is set to 204.  Some issues have been opened about this.

Depending on one's point of view, the JSONResponse 204 thing is either a bug or a gotcha.  Let's suppose that it is the latter. Then a good resolution would be to document the behavior, and possibly raise an exception when one attempts to create a JSONResponse with status code 204 (or 304).

The point of introducing an EmptyResponse class was to make explicit the concept of a response with no message body.  This is an application of a well-known software pattern. But from an ease-of-use perspective, it probably makes more sense to ensure that a Response object with status code 204 generates a correct message, since 204 is used relatively infrequently.

In general, a response should have a Content-Length header.  So perhaps the default value of the content parameter in Response.__init__ should be b'' instead of None.  Then when content is None, no content-length header is set.  

Then, the behavior of Response(status_code=204, content=None) is made explicit, while still retaining the possibility for users to do the wrong thing either with a Response or any of its subclasses.

**Comment by TBBle** [comment]
> Why would you want to add a new response class when an already existing Response can perfectly be empty?

To be _explicit_ about my intent to not return a body. If I declare _any_ response with an `EmptyResponse`, then I'm saying "If I sent data out of this, even by accident, don't send a response body". I might expect "And log an error or fail or something", but that's not as strong an expectation as "Don't let me send the thing I specified I wouldn't send".

In effect, `EmptyResponse` is like a `HEAD` request, in that I am allowed to do whatever I want in the handler, but it won't then send a body back to the client. That may actually be a good use-case for this, if I have an API that acts like "HEAD" for another API, i.e. does the same work, but only sends back the headers of the resulting response. Then I could share the handler, knowing that my 204-responding end-point won't accidentally send back an object.

A similar use-case is when I have a method which can either return a body, or return an empty response with "Your cached object is already correct". This is [a canonical use of the 304 result code](https://httpwg.org/specs/rfc7232.html#header.if-none-match) (not tested nor fully-fleshed-out, but illustrating the use-case):

```python
@app.get("/", responses={304: {description="Client copy is already correct"}):
async def read_unless_client_already_has_it(if_none_match = Header(None), response: Response):
   correctETag = check_for_existing_etag(if_none_match)
   if corectETag is not None:
     # Possible today: Has no content by absence of `content` parameter.
     return Response(status_code=304, headers=dict("E-Tag" = correctETag))

     # If we had EmptyResponse: Has no content by explicit type
     return EmptyResponse(status_code=304, headers=dict("E-Tag" = correctETag))

     # Irrespective of EmptyResponse, what we actually want from this ticket:
     # I think this is much nicer, and aesthetically matches the "slow" path below.
     response.status_code = 304
     response.headers["E-Tag"] = correctETag
     return

   content, etag = await fetch_object_from_storage()
   response.headers["E-Tag"] = etag
   return content
```

`Response` simply says "There may be some response data", and the fact it must be empty in a 204 result_code is _implicit_ and left to the user to get right.

To reflect the question, why would I want a `JSONResponse` class when already existing Response can carry perfectly JSON content with a `Content-Type` header?

This is slightly separate from the actual ticket here, which could be phrased as "`status_code=HTTP_204_NO_CONTENT` APIs with no `response_class` specified should act like the `response_class=EmptyResponse`", but to be clear, for this ticket, I thought `EmptyResponse` was an elegant solution that allowed a bunch of other stuff as I described above, but I'm not super-tied to it, as "act like the `response_class=EmptyResponse`" can very easily just be "No body data, no Content-Length header, no Transfer-Encoding header", [handled magically by Starlette](https://github.com/encode/starlette/pull/1397).

> But the scenario you have described illustrates the intention to construct the invalid response. When you use 204, you know there should be no content, so you shouldn't put any content to the response.

I have a different perspective on this.

```python
@app.get("/", status_code=HTTP_204_NO_CONTENT)
def read_root():
    print("d")
```
does not, to me, illustrate an intention to construct an invalid response. I think the intent is pretty clear: Print a value to stdout, and _return nothing_.

But it's possible you're talking about a different scenario, so I may have misunderstood.

I would agree that the below illustrates an _attempt_ to construct an invalid response, but is more-likely (and particularly in real-world usage) illustrating an _accident_ or oversight than an _intention_:

```python
@app.get("/", status_code=HTTP_204_NO_CONTENT)
def read_root():
    print("d")
    return True
```

That's why I'd be fine if that sort of code actually failed, rejecting attempts to create an invalid response (both accidental and intentional) and allowing that be fixed trivially early, but I have mentioned use-cases above where it might be desirable for code-structure if the framework ignores the return content, ensuring a valid response that suits the declaration (`status_code=HTTP_204_NO_CONTENT`).

> I don't know, maybe Starlette, after fixing headers issue, could also strip the content from the response in case it was provided with the status codes that don't allow it?

I'd be fine with that too, but I'm not _too_ fussed, because technically as long as the Content-Length and Transfer-Encoding headers are not included, it's valid HTTP, and the rendered body being included is just line-noise, which is at-mostly likely to lead to a premature disconnection of the HTTP connection. When pipelining requests, this would lead to failures in still-outstanding requests when the connection is dropped.

And this sort of issue is very possible to never hit during testing, but then encounter in the wild e.g., when someone uses a previously-unseen client library with a different behaviour in this (unspecified) case.

So it's not HTTP 1.1-spec-mandatory to strip the body, but I'd recommend it. (I'm not sure if this changed in HTTP 2, I don't know it that well. I suspect the change in framing means it'd be an error or ignored instead of being line-noise)

Of course, any lower-level libraries may barf on that body content, but that's an ASGI host implementation question, I think.

**Comment by tomchristie** [comment]
> Separately, I'm mildly concerned about the proposed change in Starlette

Welcome to expand on this. Which specific change, and what behaviour do you want/expect to see instead?

**Comment by TBBle** [comment]
The "proposed" change (now merge) was https://github.com/encode/starlette/pull/1397, and my concerns were described in that paragraph.

The tl;dr is that the given change is a HTTP syntax change which doesn't prevent invalid HTTP on-the-wire, and I'd have preferred a change that either rejected or fixed such attempts instead, i.e. the following test should have a clear behaviour:

```
def test_empty_204_response_with_content(test_client_factory):
    app = Response(status_code=204, content="hi")
    # Option 1: Response is rejected as invalid
    client: TestClient = test_client_factory(app)
    response = client.get("/")
    assert "content-length" not in response.headers
    # Option 2: Invalid content was ignored, and not put on the wire.
    assert response.text == ""
    assert client.hasMoreDataOnSocket() == False
```

**Comment by lanquil** [comment]
Is this fixed by https://github.com/tiangolo/fastapi/releases/tag/0.79.0 ?

**Comment by JarroVGIT** [comment]
Yes

**Comment by tiangolo** [comment]
Thanks for the report @aviramha! :rocket: 

And thanks for the discussion here everyone! It seems this was handled in https://github.com/tiangolo/fastapi/pull/5145, available since FastAPI `0.79.0`. :cake:
